### ISSUE TITLE
KeyError tracing to optimize_dataframe_getitem

### ISSUE THREAD

**Issue by SultanOrazbayev** [dask/dask#7692]
The following code from a [StackOverflow question](https://stackoverflow.com/questions/67645088) is raising an error with dask `2021.5.0` but not with `2021.4.1`:

```python
import dask
import dask.dataframe as dd
df = dask.datasets.timeseries()
df2 = df[df.y > 0]
df2.compute()
```

Traceback:
<details>

```pytb
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-8-a8eb713bc07c> in <module>
----> 1 df2.compute()

~/miniconda/envs/dask2/lib/python3.9/site-packages/dask/base.py in compute(self, **kwargs)
    283         dask.base.compute
    284         """
--> 285         (result,) = compute(self, traverse=False, **kwargs)
    286         return result
    287

~/miniconda/envs/dask2/lib/python3.9/site-packages/dask/base.py in compute(*args, **kwargs)
    559     )
    560
--> 561     dsk = collections_to_dsk(collections, optimize_graph, **kwargs)
    562     keys, postcomputes = [], []
    563     for x in collections:

~/miniconda/envs/dask2/lib/python3.9/site-packages/dask/base.py in collections_to_dsk(collections, optimize_graph, optimizations, **kwargs)
    335         for opt, val in groups.items():
    336             dsk, keys = _extract_graph_and_keys(val)
--> 337             dsk = opt(dsk, keys, **kwargs)
    338
    339             for opt in optimizations:

~/miniconda/envs/dask2/lib/python3.9/site-packages/dask/dataframe/optimize.py in optimize(dsk, keys, **kwargs)
     20     else:
     21         # Perform Blockwise optimizations for HLG input
---> 22         dsk = optimize_dataframe_getitem(dsk, keys=keys)
     23         dsk = optimize_blockwise(dsk, keys=keys)
     24         dsk = fuse_roots(dsk, keys=keys)

~/miniconda/envs/dask2/lib/python3.9/site-packages/dask/dataframe/optimize.py in optimize_dataframe_getitem(dsk, keys)
    103         # Project columns and update blocks
    104         old = layers[k]
--> 105         new = old.project_columns(columns)[0]
    106         if new.name != old.name:
    107             columns = list(columns)

~/miniconda/envs/dask2/lib/python3.9/site-packages/dask/layers.py in project_columns(self, columns)
    941             # Apply column projection in IO function
    942             try:
--> 943                 io_func = self.io_func.project_columns(list(columns))
    944             except AttributeError:
    945                 io_func = self.io_func

~/miniconda/envs/dask2/lib/python3.9/site-packages/dask/dataframe/io/demo.py in project_columns(self, columns)
     87         func = copy.deepcopy(self)
     88         func.columns = columns
---> 89         func.dtypes = {c: self.dtypes[c] for c in columns}
     90         return func
     91

~/miniconda/envs/dask2/lib/python3.9/site-packages/dask/dataframe/io/demo.py in <dictcomp>(.0)
     87         func = copy.deepcopy(self)
     88         func.columns = columns
---> 89         func.dtypes = {c: self.dtypes[c] for c in columns}
     90         return func
     91

KeyError: 'gt-846fc0648f9e5e57119c74d81b65378c'
```
</details>

**Environment**:

- Dask version: 2021.5.0  (this error is not raised with dask 2021.4.1)
- Python version: 3.9.4
- Operating System: MacOS BigSur
- Install method (conda, pip, source): conda

**Comment by rjzamora** [comment]
Thanks for raising this @SultanOrazbayev !  I am attempting to include a fix for this in #7663 - I am also narrowing the scope of that PR, so I plan to remove the "WIP" label today.

**Comment by rjzamora** [comment]
Note - Fix was moved to #7698
